src/getValidators.ts

```diff
@@ -0,0 +1,179 @@
+  const [validators, states] = getValidatorsFromFunctionData({
+    abi: [setValidatorABI],
+    data: input,
+  });
+
+  validators.forEach((validator, i) => {
+    const isAdd = states[i];
+    if (isAdd) {
+      acc.add(validator);
+    } else {
+      acc.delete(validator);
+    }
+  });
+
+  return acc;
+}
+
+type GetValidatorsParams = {
+  /** Address of the rollup we're getting list of validators from */
+  rollup: Address;
+};
+type GetValidatorsReturnType = {
+  /**
+   * If logs contain unknown signature, validators list might:
+   * - contain false positives (validators that were removed, but returned as validator)
+   * - contain false negatives (validators that were added, but not present in the list)
+   */
+  isAccurate: boolean;
+  /** List of validators for the given rollup */
+  validators: Address[];
+};
+
+/**
+ *
+ * @param {PublicClient} publicClient - The chain Viem Public Client
+ * @param {GetValidatorsParams} GetValidatorsParams {@link GetValidatorsParams}
+ *
+ * @returns Promise<{@link GetValidatorsReturnType}>
+ *
+ * @remarks validators list is not guaranteed to be exhaustive if the `isAccurate` flag is false.
+ * It might contain false positive (validators that were removed, but returned as validator)
+ * or false negative (validators that were added, but not present in the list)
+ *
+ * @example
+ * const { isAccurate, validators } = getValidators(client, { rollup: '0xc47dacfbaa80bd9d8112f4e8069482c2a3221336' });
+ *
+ * if (isAccurate) {
+ *   // Validators were all fetched properly
+ * } else {
+ *   // Validators list is not guaranteed to be accurate
+ * }
+ */
+export async function getValidators<TChain extends Chain | undefined>(
+  publicClient: PublicClient<Transport, TChain>,
+  { rollup }: GetValidatorsParams,
+): Promise<GetValidatorsReturnType> {
+  let blockNumber: bigint | 'earliest';
+  try {
+    const createRollupTransactionHash = await createRollupFetchTransactionHash({
+      rollup,
+      publicClient,
+    });
+    const receipt = await publicClient.waitForTransactionReceipt({
+      hash: createRollupTransactionHash,
+    });
+    blockNumber = receipt.blockNumber;
+  } catch (e) {
+    blockNumber = 'earliest';
+  }
+
+  const events = await publicClient.getLogs({
+    address: rollup,
+    event: ownerFunctionCalledEventAbi,
+    args: { id: 6n },
+    fromBlock: blockNumber,
+    toBlock: 'latest',
+  });
+
+  const txs = await Promise.all(
+    events.map((event) =>
+      publicClient.getTransaction({
+        hash: event.transactionHash,
+      }),
+    ),
+  );
+
+  let isAccurate = true;
+  const validators = txs.reduce((acc, tx) => {
+    const txSelectedFunction = tx.input.slice(0, 10);
+
+    switch (txSelectedFunction) {
+      case createRollupFunctionSelector: {
+        const [{ validators }] = getValidatorsFromFunctionData({
+          abi: [createRollupABI],
+          data: tx.input,
+        });
+
+        acc = new Set([...acc, ...validators]);
+        return acc;
```

Comment:
I'd much prefer it without direct reassignment
```suggestion
        return new Set([...acc, ...validators]);
```